RFC.md

```diff
@@ -0,0 +1,253 @@
+2. Experts under specialization. Ideally the gating network will learn token-to-expert affinity such that similar or relevant tokens are routed to the same expert. However, a sub-optimal strategy can produce redundant experts and/or experts that are not sufficiently specialized.
+3. Same compute for each token. Token choice will allocate a fixed number of experts to each token regardless of the importance of different tokens. Ideally an MOE model should flexibly allocate compute resources based on the complexity of the input.
+
+Compared to **token choice**, **expert choice** topK routing lets experts select its top-k tokens. The ExpertChoiceTopKRouter class routes input tokens to different experts based on the router scores.
+
+**Here's the proposed Expert Choice Routing design in torchtune:**
+```python
+class ExpertChoiceTopKRouter(nn.Module):
+    def __init__(self, hidden_dim, num_experts):
+        self.gate = nn.Linear(hidden_dim, num_experts)
+        self.tokens_per_expert = tokens_per_expert
+
+    def forward(self, x, use_sigmoid=False):
+        '''
+        input:
+            x: shape [bs*slen, hidden_dim]
+        outputs:
+            routed_input: selected tokens
+                shape [num_experts*tokens_per_expert, hidden_dim]
+            token_indices: selected token indices
+            num_local_tokens_per_expert: None
+        '''
+        # scores shape [num_experts, bs*slen]
+        scores = self.gate(x).transpose(0,1)
+        if use_sigmoid:
+            scores = torch.sigmoid(scores.to(sigmoid_dtype)).to(x.dtype)
+        else:
+            scores = F.softmax(scores.to(softmax_dtype), dim=0).to(x.dtype)
+        # router scores/indices shape [num_experts, tokens_per_expert]
+        top_scores, selected_token_indices = torch.topk(scores, k=self.tokens_per_expert, dim=1)
+
+        # apply the token preprocess function and then run experts forward
+        token_indices = selected_token_indices.reshape(-1, 1).expand(-1, D)
+        # routed input shape [num_experts*tokens_per_expert, hidden_dim]
+        routed_input = torch.gather(x, dim=0, index=token_indices)
+        routed_input = routed_input * top_scores.reshape(-1, 1)
+        return routed_input, token_indices, None,
+```
+
+## Moe Layer
+An MOE layer consists of experts and routers.
+
+**Here's the proposed MoeLayer design in torchtune:**
+```python
+class MoeLayer(nn.Module):
+    def __init__(self, router="token_choice"):
+        self.experts = moe_experts(hidden_dim, model_dim, num_experts=num_experts)
+        self.shared_expert = moe_expert(hidden_dim, model_dim)
+        if router == "token_choice":
+            self.router = TokenChoiceTopKRouter(hidden_dim, num_experts, experts_per_token)
+        elif router == "expert_choice":
+            self.router = ExpertChoiceTopKRouter(hidden_dim, num_experts, tokens_per_expert)
+        else:
+            raise NotImplementedError("This router is not supported yet!")
```

Comment:
This is a minor point, but I wouldn't configure router type using strings. Just pass as an nn.Module as that's more generic (then anyone with their own custom router matching the same signature can use out of the box)